build.sbt

```diff
@@ -1,45 +1,76 @@
+import build.BuildType
+
 name := "lakefs-spark-client"
-scalaVersion := "2.12.10"
-
-lazy val core = (project in file("core"))
-  .settings(
-    Compile / PB.includePaths += (Compile / resourceDirectory).value,
-    Compile / PB.protoSources += (Compile / resourceDirectory).value,
-    Compile / PB.targets := Seq(
-      scalapb.gen() -> (Compile / sourceManaged).value
-    ),
-    sharedSettings,
-  )
-  .settings(fatPublishSettings)
-lazy val examples = (project in file("examples")).dependsOn(core)
-  .settings(
-    sharedSettings,
-  )
-  .settings(
-    mainClass in assembly := Some("io.treeverse.examples.List"),
-  )
-  .settings(fatPublishSettings)
 
-// Use an older JDK to be Spark compatible
-javacOptions ++= Seq("-source", "1.8", "-target", "1.8")
-scalacOptions ++= Seq("-release", "8", "-target:jvm-1.8")
+lazy val projectVersion = "0.1.0-SNAPSHOT"
 
-core / libraryDependencies ++= Seq("org.rocksdb" % "rocksdbjni" % "6.6.4",
-  "commons-codec" % "commons-codec" % "1.15",
-  "org.apache.spark" %% "spark-sql" % "3.0.1" % "provided",
-  "com.thesamet.scalapb" %% "scalapb-runtime" % scalapb.compiler.Version.scalapbVersion % "protobuf",
-  "org.apache.hadoop" % "hadoop-aws" % "2.7.7",
-  "org.apache.hadoop" % "hadoop-common" % "2.7.7",
-  "org.scalaj" %% "scalaj-http" % "2.4.2",
-  "org.json4s" %% "json4s-native" % "3.7.0-M8",
-  "com.google.guava" % "guava" % "16.0.1",
-  "com.google.guava" % "failureaccess" % "1.0.1",
-)
+// Spark versions 2.4.7 and 3.0.1 use different Scala versions.  Changing this is a deep
+// change, so key the Spark distinction by the Scala distinction.  sbt doesn't appear to
+// support other ways of changing emitted Scala binary versions using the same compiler,
+// sigh...
 
-examples / libraryDependencies ++= Seq(
-  "org.apache.spark" %% "spark-sql" % "3.0.1" % "provided",
+// SO https://stackoverflow.com/a/60177627/192263 hints that we cannot use 2.11 here before
+// this version
+lazy val scala211Version = "2.11.12"
+lazy val scala212Version = "2.12.12"
+
+lazy val commonSettings = Seq(
+  /*
+   * Define the scala sources relative to the sub-directory the project source. The
+   * individual projects have their sources defined in `./target/${projectName}`,
+   * therefore `./src` lives two directories above base. Also do this for `resources`
+   * etc, if needed.
+   */
+  scalaSource in Compile := baseDirectory.value / ".." / ".." / "core" / "src" / "main" / "scala",
+  scalaSource in Test := baseDirectory.value / ".." / ".." / "core" / "src" / "test" / "scala",
 )
 
+def generateProject(buildType: BuildType) = {
+  lazy val core = Project(s"core-${buildType.name}", file(s"target/core:${buildType.name}"))
+    .settings(
+      sharedSettings,
+      scalaVersion := buildType.scalaVersion,
+      Compile / resourceDirectory := baseDirectory.value / ".." / ".." / "core" / "src" / "main" / "resources",
```

Comment:
No, because `examples` gets different settings.  We _could_ try to generate all paths using a function, and then call a that function with different args in `core` and in `examples`.  It will avoid repeating configuration, but may make the SBT more like a program and therefore less explicit.  Pushing as a separate commit, will do whichever you prefer.